### PR TITLE
Batchsizetask

### PR DESCRIPTION
**Pull request by Darktex**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
*

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [ ] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [ ] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 1]
```diff
@@ -537,81 +620,104 @@ def _log_metrics(
...
+                )
+        padded_prompt_tokens = pad_and_stack(
+            prompt_tokens, pad_value=self._tokenizer.pad_id
+        )
+        padded_response_tokens = pad_and_stack(
+            response_tokens, pad_value=self._tokenizer.pad_id
+        )
+        padded_logprobs = pad_and_stack(logprobs, pad_value=0.0)
+
+        # TODO if the below is uncommented, VLLM doesn't start. The error is masked
```
Comment (Darktex) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:678]: I'm going crazy with this one. The fact that the error is getting masked is making this tough to debug... Has anyone else seen this??

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 2]
```diff
@@ -62,26 +66,61 @@
...
+import torch.nn.functional as F
+
+
+def pad_and_stack(tensors, pad_value=0):
+    """
+    Pads a list of tensors to the same shape and stacks them.
+
+    Args:
+        tensors (list of torch.Tensor): List of tensors to pad and stack.
+        pad_value (scalar): The value to pad with (default is 0).
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:100]: we prob want to use list[torch.Tensor[ and int, instead of scalar. If you are using precommit-hooks, it will complain about not having the type hints in the args too

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 3]
```diff
@@ -62,26 +66,61 @@
...
+        # Build padding amounts for each dimension (pad on the right only).
+        # F.pad expects the padding tuple in reverse order: (pad_left_last_dim, pad_right_last_dim, ...).
+        pad = []
+        for dim in range(num_dims - 1, -1, -1):
+            pad_amount = max_shape[dim] - tensor.size(dim)
+            pad.extend([0, pad_amount])
+        padded = F.pad(tensor, pad, value=pad_value)
+        padded_tensors.append(padded)
+
+    return torch.stack(padded_tensors)
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:122]: preallocating a tensor should be more efficient. Something like below (didnt test)

```
def pad_and_stack(tensors: List[torch.Tensor], pad_value:int=0):
    """
    Pads a list of tensors to the same shape and stacks them efficiently.
    
    Args:
        tensors (List[torch.Tensor]): List of tensors to pad and stack.
        pad_value (int): The value to pad with (default is 0).
    
    Returns:
        torch.Tensor: A tensor with shape (N, *max_dims) where each tensor is padded to max_dims.
    """
    # Check if the tensor list is empty
    if not tensors:
        raise ValueError("The list of tensors cannot be empty.")
    
    # Assume all tensors have the same number of dimensions
    num_dims = tensors[0].dim()
    
    # Compute the maximum size for each dimension
    max_shape = [max(t.size(d) for t in tensors) for d in range(num_dims)]
    
    # Number of tensors
    N = len(tensors)
    
    # Preallocate the result tensor filled with pad_value
    result = torch.full(
        (N, *max_shape),
        pad_value,
        dtype=tensors[0].dtype,
        device=tensors[0].device
    )
    
    # Assign each tensor to the corresponding slice
    for i, tensor in enumerate(tensors):
        # Create a tuple of slices based on the tensor's shape
        slices = tuple(slice(0, s) for s in tensor.shape)
        result[i][slices] = tensor
    
    return result
    ````

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 4]
```diff
@@ -94,6 +133,54 @@ class Trajectory(TensorClass["nocast"]):
...
+class PaddedTrajectoryBatch(TensorClass["nocast"]):
+    """
+    These are only single-turn trajectories (prompt + response).
+    TODO design together what a multi-turn interaction looks like.
+    Not needed at this stage.
+    """
+
+    policy_model_id: int  # Fingerprints the policy model who wrote the response
+    prompts_str: str  # size is [batch]
+    responses_str: str  # size is [batch, grpo_size]
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:145]: I am a big confused about the 'str' and the shape. Do we need to do json.loads(prompts_str)? 

If so, should it just be a List[str] instead?

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 5]
```diff
@@ -94,6 +133,54 @@ class Trajectory(TensorClass["nocast"]):
...
 
 
+class PaddedTrajectoryBatch(TensorClass["nocast"]):
+    """
+    These are only single-turn trajectories (prompt + response).
+    TODO design together what a multi-turn interaction looks like.
+    Not needed at this stage.
+    """
+
+    policy_model_id: int  # Fingerprints the policy model who wrote the response
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:143]: maybe this should go under metadata

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 6]
```diff
@@ -537,81 +633,52 @@ def _log_metrics(
...
+            prompts_str=prompts_str,
+            responses_str=responses_str,
+            prompt_tokens=padded_prompt_tokens,
+            response_tokens=padded_response_tokens,
+            logprobs=padded_logprobs,
+            # Initialize ref_logprobs as zeros - will be filled by RefActor
+            ref_logprobs=torch.zeros_like(padded_logprobs),
+            batch_size=batch_size,
+        )
+        return
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:664]: is return None intentional?

File: scripts/runnable_recipe_ray_vllm_weight_sync.py [hunk 7]
```diff
@@ -1745,7 +1836,11 @@ def setup(self, cfg):
             actor_options={"num_cpus": 10, "num_gpus": 0},
             maxsize=cfg.vllm.queue_maxsize,
         )
-        self.replay_buffer = RayReplayBuffer(storage=functools.partial(LazyStackStorage, max_size=1000), batch_size=1, remote_config={"num_cpus": 10, "num_gpus": 0})
+        self.replay_buffer = RayReplayBuffer(
+            storage=functools.partial(LazyStackStorage, max_size=1000),
```
Comment (felipemello1) [inline scripts/runnable_recipe_ray_vllm_weight_sync.py:1840]: in main i changed max_size=3. If you keep as 1000, rewards wont go up.

**Comment by Darktex** [comment]
Implemented in #24
